haystack/components/generators/chat/hugging_face_local.py

```diff
@@ -362,6 +409,37 @@ def _validate_stop_words(self, stop_words: Optional[List[str]]) -> Optional[List
             )
             return None
 
-        # deduplicate stop words
-        stop_words = list(set(stop_words or []))
-        return stop_words
+        return list(set(stop_words or []))
+
+    def _parse_tool_call(self, text: str) -> Optional[List[ToolCall]]:
+        """
+        Parse a tool call from model output text.
+
+        :param text: The text to parse for tool calls.
+        :returns: A ToolCall object if a valid tool call is found, None otherwise.
+        """
+        # if the tool pattern is a callable, call it with the text and return the result
+        if callable(self.tool_pattern):
+            return self.tool_pattern(text)
+
+        # if the tool pattern is a regex pattern, search for it in the text
+        try:
+            match = re.search(self.tool_pattern, text, re.DOTALL)
+        except re.error:
+            logger.warning("Invalid regex pattern for tool parsing: {pattern}", pattern=self.tool_pattern)
+            return None
+
+        if not match:
+            return None
+
+        # seem like most models are not producing tool ids, so we omit them
+        # and just use the tool name and arguments
+        name = match.group(1) or match.group(3)
+        args_str = match.group(2) or match.group(4)
```

Comment:
this part makes me think that this parsing method works only with a specific regexp (very similar to `DEFAULT_TOOL_PATTERN`).

So I would like to propose the following idea:
- `tool_pattern` can become `tool_parsing_function`, only accepting callables and defaulting to a standard implementation
- the standard implementation is just a function which does tool parsing using `DEFAULT_TOOL_PATTERN`

Clear? Do you see any drawbacks with this approach?

Reply:
Great idea!